File: src/helperFunctions.ts
import { JSDOM } from 'jsdom';
import { UserData } from './types';

export function parseJsonCookies(json: any) {
  const cookieEntries = normalizeCookieEntries(json);

  const output = {
    geo: "",
    hcmex: "",
    deviceId: "",
    domain: "",
    expires: 0
  }

  for (const item of cookieEntries) {
    const normalized = normalizeCookieItem(item);
    if (!normalized) continue;

    switch (normalized.name) {
      case "geo":
        output.geo = normalized.value
        break;
      case "_hcmex_key":
        output.hcmex = normalized.value
        output.domain = normalized.domain || output.domain
        output.expires = normalizeExpiration(normalized.expires)
        break;
      case "device_id":
        output.deviceId = normalized.value
        break;
    }
  }

  return output
}

function normalizeCookieEntries(json: any): any[] {
  if (Array.isArray(json)) return json;
  if (json && Array.isArray(json.cookies)) return json.cookies;
  if (json && json.data && Array.isArray(json.data.cookies)) return json.data.cookies;
  if (json && json.cookieHeader && typeof json.cookieHeader === 'string') return parseCookieHeader(json.cookieHeader);
  if (json && json.cookie && typeof json.cookie === 'string') return parseCookieHeader(json.cookie);
  if (json && json.headers && typeof json.headers.cookie === 'string') return parseCookieHeader(json.headers.cookie);
  if (json && typeof json.Cookie === 'string') return parseCookieHeader(json.Cookie);
  if (json && isPlainObject(json)) {
    const hasLikelyCookieKeys = ['geo', '_hcmex_key', 'device_id'].some((k) => k in json);
    if (hasLikelyCookieKeys) {
      return Object.entries(json).map(([name, value]) => ({
        name,
        value,
        domain: json.domain ?? json.host ?? '',
        expires: json.expirationDate ?? json.expires ?? json.expiry ?? 0
      }));
    }
  }
  return [];
}

function normalizeCookieItem(item: any): { name: string; value: string; domain: string; expires: unknown } | null {
  if (!item || typeof item !== 'object') return null;
  const name = String(item.name ?? item.key ?? item.cookieName ?? item.cookie ?? '');
  const value = item.value ?? item.content ?? item.val ?? item.cookieValue;
  if (!name || value === undefined || value === null) return null;

  return {
    name,
    value: String(value),
    domain: String(item.domain ?? item.host ?? item.hostname ?? ''),
    expires: item.expirationDate ?? item.expires ?? item.expiry ?? item.expiresAt ?? 0
  };
}

function parseCookieHeader(cookieHeader: string): Array<{ name: string; value: string; domain: string; expires: number }> {
  return cookieHeader
    .split(';')
    .map((part) => part.trim())
    .filter(Boolean)
    .map((part) => {
      const idx = part.indexOf('=');
      const name = idx >= 0 ? part.slice(0, idx).trim() : part.trim();
      const value = idx >= 0 ? part.slice(idx + 1).trim() : '';
      return { name, value, domain: '', expires: 0 };
    })
    .filter((item) => item.name.length > 0);
}

function normalizeExpiration(raw: unknown): number {
  const value = Number(raw);
  if (!Number.isFinite(value) || value <= 0) return 0;
  // Some exports use seconds, others milliseconds.
  return value > 1e12 ? value : value * 1000;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}



export async function getCsrfTokes(data: UserData) {
  const MINIMAL_REQUEST = {
    // Only these 3 cookies are actually needed
    headers: {
      'Cookie': [
        '_hcmex_key=' + data.cookies.hcmex,
        'device_id=' + data.cookies.deviceId,
        'geo=' + data.cookies.geo
      ].join('; ')
    }
  };

  const response = await fetch(`https://${data.cookies.domain}/`, MINIMAL_REQUEST);
  const text = await response.text();
  const dom = new JSDOM(text)
  const document = dom.window.document
  const metaCsrf = document.querySelector('meta[name="csrf"]')?.getAttribute('content') || null;

  const chronoResponse = await fetch(`https://${data.cookies.domain}/chrono/${data.userId}/hub_chrono`, MINIMAL_REQUEST);
  const chronoText = await chronoResponse.text();
  const chronoDom = new JSDOM(chronoText)
  const chronoDocument = chronoDom.window.document;
  const inputCsrf = chronoDocument.querySelector('input[name="_csrf_token"]')?.getAttribute("value") || null;

  return { metaCsrf, inputCsrf }
}

File: rust-src/helper_functions.rs
use anyhow::{anyhow, Result};
use reqwest::Client;
use scraper::{Html, Selector};
use serde_json::Value;

use crate::types::{Cookies, UserData};

pub fn parse_json_cookies(json: &Value) -> Result<Cookies> {
    let arr = json
        .as_array()
        .ok_or_else(|| anyhow!("Expected cookie JSON array"))?;

    let mut geo = String::new();
    let mut hcmex = String::new();
    let mut device_id = String::new();
    let mut domain = String::new();
    let mut expires = 0_i64;

    for item in arr {
        let Some(name) = item.get("name").and_then(Value::as_str) else {
            continue;
        };

        match name {
            "geo" => {
                geo = item
                    .get("value")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string();
            }
            "_hcmex_key" => {
                hcmex = item
                    .get("value")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string();
                domain = item
                    .get("domain")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string();
                let exp = item
                    .get("expirationDate")
                    .and_then(Value::as_f64)
                    .unwrap_or_default();
                expires = (exp * 1000.0) as i64;
            }
            "device_id" => {
                device_id = item
                    .get("value")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string();
            }
            _ => {}
        }
    }

    Ok(Cookies {
        geo,
        hcmex,
        device_id,
        domain,
        expires,
    })
}

pub async fn get_csrf_tokens(client: &Client, data: &UserData) -> Result<(Option<String>, Option<String>)> {
    let cookie_header = format!(
        "_hcmex_key={}; device_id={}; geo={}",
        data.cookies.hcmex, data.cookies.device_id, data.cookies.geo
    );

    let home_url = format!("https://{}/", data.cookies.domain);
    let home_html = client
        .get(home_url)
        .header("Cookie", &cookie_header)
        .send()
        .await?
        .text()
        .await?;

    let meta_csrf = extract_selector_attr(&home_html, "meta[name=\"csrf\"]", "content");

    let chrono_url = format!(
        "https://{}/chrono/{}/hub_chrono",
        data.cookies.domain, data.user_id
    );
    let chrono_html = client
        .get(chrono_url)
        .header("Cookie", &cookie_header)
        .send()
        .await?
        .text()
        .await?;

    let input_csrf = extract_selector_attr(&chrono_html, "input[name=\"_csrf_token\"]", "value");

    Ok((meta_csrf, input_csrf))
}

fn extract_selector_attr(html: &str, selector: &str, attr: &str) -> Option<String> {
    let doc = Html::parse_document(html);
    let sel = Selector::parse(selector).ok()?;
    doc.select(&sel)
        .next()
        .and_then(|el| el.value().attr(attr))
        .map(|s| s.to_string())
}
